Improved check against integer overflow, including more borderline te…

heplesser opened the pull request:
This PR should fix #36 by adding safe overflow tests.
@apeyser @tammoippen

tammoippen commented:
#39 Fixed the TravisCI .yaml. To rerun the TravisCI build on the most current source state, you have to do another commit in the heplesser:fix_36_test_rdv_param_setting_fails branch, e.g.

    git commit --amend   # this just touchs the last commit and does not modify anything
    git push origin fix_36_test_rdv_param_setting_fails

tammoippen commented:
The code looks very nice to me. Thank you for the article on integer overflow. +1 for merge after TravisCI is green.

abigailm commented:
cpp looks fine to me but I can't comment on the SLI, so a +1 with caveat.

gewaltig commented on testsuite/unittests/test_rdv_param_setting.sli:
```diff
@@ -244,20 +244,39 @@ statusdict /have_gsl :: {
   /MT19937 /uniform_int << /low 11 /high 10 >> run_test
 } fail_or_die
 
+/:range_limit
```
I am not sure I follow what this function does. 
Can you maybe comment the parameter?

gewaltig replied:
@heplesser My problem is the function :range_limit 
Can you add a comment, explaining what parameter it expects and what it is exactly computing in an overflow-safe way?

gewaltig commented on testsuite/unittests/test_rdv_param_setting.sli:
```diff
@@ -244,20 +244,39 @@ statusdict /have_gsl :: {
   statusdict/architecture/long :: 8 mul   % N
   2 sub 2 exch pow cvi                    % 2^(N-2)
   dup 1 sub % stack: 2^(N-2) 2^(N-2)-1
   add       % 2^(N-1)-1
```
Calculation should be fine.

heplesser commented:
@gewaltig I just pushed an updated version with a few more explanations.

heplesser commented:
@gewaltig Could you check whether this PR is okay with the explanations I added on Monday and then give it a +1, so that the PR could be merged? Thank you!

gewaltig commented:
Will do in 30 minutes.
Marc-Oliver

P.S. sent from a mobile device with small keyboard. Please excuse brevity and typos.

Am 24.06.2015 um 17:10 schrieb Hans Ekkehard Plesser <notifications@github.com<mailto:notifications@github.com>>:


@gewaltig<https://github.com/gewaltig> Could you check whether this PR is okay with the explanations I added on Monday and then give it a +1, so that the PR could be merged? Thank you!

—
Reply to this email directly or view it on GitHub<https://github.com/nest/nest-simulator/pull/37#issuecomment-114901629>.

heplesser commented:
@gewaltig I have now updated the code. The function always computed the same value, so instead of running the function, I just define a constant now. And I call that by its usual name for the value it stores, `:LONG_MAX`. I also added some more explanation of the way the value is computed.

heplesser commented:
@gewaltig @lekshmideepu The Travis failure seems to be due to some technical trouble at Travis, all build variants failed after just 10 seconds or so. I will push a trivial change to trigger Travis once more.

gewaltig commented:
SLI code looks good to me now. +1 for merging
